cranelift/codegen/src/isa/riscv64/inst/encode.rs

```diff
@@ -27,6 +27,21 @@ pub fn encode_r_type(opcode: u32, rd: Reg, funct3: u32, rs1: Reg, rs2: Reg, func
     bits
 }
 
+/// Encode an I-type instruction.
+///
+/// Layout:
+/// 0-------6-7-------11-12------14-15------19-20------------------31
+/// | Opcode |   rd     |  width   |   rs1    |     Offset[11:0]    |
+pub fn encode_i_type(opcode: u32, rd: Reg, width: u32, rs1: Reg, offset: Imm12) -> u32 {
+    let mut bits = 0;
+    bits |= opcode & 0b1111111;
+    bits |= reg_to_gpr_num(rd) << 7;
+    bits |= (width & 0b111) << 12;
+    bits |= reg_to_gpr_num(rs1) << 15;
+    bits |= (offset.as_u32() & 0b1111_1111_1111) << 20;
```

Comment:
This turned out really neatly, I've implemented these suggestions in https://github.com/bytecodealliance/wasmtime/pull/6325/commits/fb87186256527cfe5f6cd0301162e9583c9955e7 since Alex had some similar concerns. I'm going to get that PR merged, and do something similar for these.